### PR TITLE
Optimize away add request overheads in dummy ep cuda-graphed forward passes

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
When running a cuda-graphed dummy forward pass on ep ranks with no work, we resort to calling add_requests to create dummy inputs. This is extremely costly on GB200s. This PR optimizes this codepath, by only adjusting the total request count and query request lengths. Decode step times for a single request (EP=4) on GB200s, drop from 30ms -> 20ms!

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/contexts/dynamic_context.py [hunk 1]
```diff
@@ -1294,7 +1306,28 @@ def initialize_attention_state(
...
+                    + smallest_cuda_graph_dimensions.decode_req_count
+                )
+                # 2. Reset request query lengths
+                # Zero out request_query_lengths to prevent stale data from causing
+                # out of bounds memory accesses in last_token_logits.
+                # This is needed because when we move finished requests to the right, we never
+                # zero out their request lengths.
+                self.request_query_lengths[0 : self.total_request_count].fill_(0)
         else:
+            if is_expert_parallel_dummy_cuda_graph_step:
```
Comment (kvareddy) [inline megatron/core/inference/contexts/dynamic_context.py]: This does not seem correct, same condition in both if and else.

Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py]: @sidsingh-nvidia It might be cleaner to handle the `is_expert_parallel_dummy_cuda_graph_step` case in a separate function

File: megatron/core/inference/contexts/dynamic_context.py [hunk 2]
```diff
@@ -1294,7 +1306,28 @@ def initialize_attention_state(
...
+                self.total_request_count = (
+                    smallest_cuda_graph_dimensions.prefill_req_count
+                    + smallest_cuda_graph_dimensions.decode_req_count
+                )
+                # 2. Reset request query lengths
+                # Zero out request_query_lengths to prevent stale data from causing
+                # out of bounds memory accesses in last_token_logits.
+                # This is needed because when we move finished requests to the right, we never
+                # zero out their request lengths.
+                self.request_query_lengths[0 : self.total_request_count].fill_(0)
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: Can we verify that this is safe / has the intended behavior for `last_token_logits` with a unit test?

File: megatron/core/inference/contexts/dynamic_context.py [hunk 3]
```diff
@@ -1261,26 +1261,38 @@ def num_decode_requests(self) -> int:
...
 
-        batch_dimensions = InferenceBatchDimensions(
-            token_count=self.active_token_count,
-            prefill_req_count=self.num_prefill_requests,
-            decode_req_count=self.num_decode_requests,
-        )
+        if is_expert_parallel_dummy_cuda_graph_step:
+            # attempt to use the smallest possible cuda graph for the dummy forward
+            smallest_cuda_graph_dimensions = min(self.cuda_graph_batch_dimensions_list)
+            batch_dimensions = smallest_cuda_graph_dimensions
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: Is this compatible with `strict` CUDA graph matching?

File: megatron/core/inference/contexts/dynamic_context.py (hunk 2 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: update this to something saner

Comment (sidsingh-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: clarify that this is handling the real config.

Comment (sidsingh-nvidia) [reply megatron/core/inference/contexts/dynamic_context.py]: move all the dummy logic to a separate function

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 4]
```diff
@@ -846,8 +849,7 @@ def dummy_forward(self):
         # a dummy cuda graph.
         input_ids, position_ids = self._dynamic_step_context_init(
             # try to use the smallest cuda-graph config for dummy forward
```
Comment (janEbert) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: This comment doesn't make sense anymore with the removal of the line below it, right?

File: megatron/inference/utils.py [hunk 5]
```diff
@@ -303,7 +303,7 @@ def get_inference_config_from_model_and_args(model: MegatronModule, args):
         track_paused_request_events=args.inference_dynamic_batching_track_paused_request_events,
         enable_chunked_prefill=args.enable_chunked_prefill,
         metrics_writer=metrics_writer,
-        logging_step_interval=args.inference_logging_step_interval,
+        logging_step_interval=args.inference_logging_step_interval
```
Comment (janEbert) [inline megatron/inference/utils.py]: This seems like something `tools/autoformat.sh` wouldn't let pass. ;)

File: megatron/core/inference/contexts/dynamic_context.py [hunk 6]
```diff
@@ -1260,38 +1261,101 @@ def num_decode_requests(self) -> int:
...
+            self.is_creating_cuda_graphs and is_expert_parallel_dummy_cuda_graph_step
+        ), "Dummy expert model parallel steps should not be creating cuda graphs."
 
         # If in CUDA graph creation mode, add dummy requests for CUDA graph capture
-        if self.is_creating_cuda_graphs:
-            self.add_dummy_requests_for_cudagraph_capture(construct_graph_dimensions)
+        if is_expert_parallel_dummy_cuda_graph_step:
+            self.add_dummy_requests_for_expert_parallel_step()
+        else:
+            if self.is_creating_cuda_graphs:
```
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py]: nit: `elif self.is_creating_cuda_graphs:`

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 4 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py]: Thanks for catching this. Fixed!

File: megatron/inference/utils.py (hunk 5 above)
Comment (sidsingh-nvidia) [reply megatron/inference/utils.py]: fixed, thanks!

**Comment by sidsingh-nvidia** [comment]
@santhnm2 I have added the unit test.
